private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/remote/Downloader.java

```diff
@@ -0,0 +1,222 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.logging.Logger;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+import org.w3c.dom.Document;
+import org.xml.sax.SAXException;
+
+public class Downloader {
+
+  private static final Logger LOG = Logger.getLogger(Downloader.class.getName());
+  // Lifted from RealisedMavenModuleResolveMetadata in Gradle's source
+  private static final Set<String> JAR_PACKAGINGS =
+      Set.of("jar", "ejb", "bundle", "maven-plugin", "eclipse-plugin");
+  // Packaging types (extensions) we know don't have fallbacks. I know `jar` is in both sets.
+  private static final Set<String> NO_FALLBACK_PACKAGINGS =
+      Set.of("jar", "tar.gz", "tar.bz2", "tar", "zip", "exe", "dll", "so");
+  private static final Set<String> NO_FALLBACK_CLASSIFIERS = Set.of("sources", "javadoc");
+  private final Path localRepository;
+  private final Set<URI> repos;
+  private final boolean cacheDownloads;
+  private final HttpDownloader httpDownloader;
+
+  public Downloader(
+      Netrc netrc,
+      Path localRepository,
+      Collection<URI> repositories,
+      EventListener listener,
+      boolean cacheDownloads) {
+    this.localRepository = localRepository;
+    this.repos = Set.copyOf(repositories);
+    this.cacheDownloads = cacheDownloads;
+    this.httpDownloader = new HttpDownloader(netrc, listener);
+  }
+
+  public DownloadResult download(Coordinates coords) {
+    DownloadResult result = performDownload(coords);
+    if (result != null) {
+      return result;
+    }
+
+    // There is a fallback we can use.
+    // RealisedMavenModuleResolveMetadata.getArtifactsForConfiguration
+    // says that if the artifact isn't a "known java packaging", then we should just look for a
+    // `jar`
+    // variant as well.
+    if (isFallbackAvailable(coords)) {
+      result = performDownload(coords.setExtension("jar"));
+    }
+
+    if (result != null) {
+      return result;
+    }
+
+    // Are we dealing with a packaging dep? Download the `pom.xml` and check
+    String originalTarget = new MavenRepositoryPath(coords).getPath();
+    String pomName = String.format("%s-%s.pom", coords.getArtifactId(), coords.getVersion());
+    String pom = Paths.get(originalTarget).getParent().resolve(pomName).toString();
+
+    DownloadResult pomResult = performDownload(coords, pom);
+    if (pomResult == null) {
+      System.out.println("\n[WARNING] The POM for " + coords + " is missing, no dependency information available");
```

Comment:
Intended to match the output by maven for the same error, we can change if needed